### DISCUSSION TITLE
How to build and debug git-gui?

### DISCUSSION THREAD

**Discussion by ToBoMi** [git-for-windows/git#5061] [Q&A]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit? 64bit

```
$ git --version --build-options

git version 2.45.2.GIT
cpu: AMD64
no commit associated with this build
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19045.4529]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
> type "$env:USERPROFILE\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt


```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

I build using Visual Studio and git-for-windows opened as a folder (As mentioned in the Readme).
Confugration and Build succeeds - git.exe starts.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

I don't really know. I use the Visual Studio Debugger/Launcher - maybe it is just `cmd`.

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```

```
 - What did you expect to occur after running these commands?

Not applicable - I have a question about building.

 - What actually happened instead?

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

# My question

I would like to contribute to git-gui and therefore want to build it. It doesn't necessarily needs to be build since it is tcl script based, but I cannot find a tcl script for git gui - like for gitk.
I also saw that the git installation insalled by using the installer of git-for-windows comes with a `git-gui.exe` (not tcl-only) and the [readme in the git-gui folder](https://github.com/git-for-windows/git/tree/main/git-gui#building-and-installing) mentions a "build" using `make`. This seems to be not really needed, but can be done.
Since gitk works without build (`tcl`) I wonder how `git-gui` is build here. I found no specific instructions for `git-gui` and I don't really know how to use `make` for git-for-windows.
I can build git (`git.exe`) and run gitk (`tclsh.exe .\\gitk-git\\gitk`)

Can someone help me start debugging of git-gui?

**Comment by dscho** [comment]
> It doesn't necessarily needs to be build since it is tcl script based, but I cannot find a tcl script for git gui - like for gitk.

It is here: https://github.com/git-for-windows/git/blob/HEAD/git-gui/git-gui.sh

But that is not all there is to it, it comes with a library of Tcl files at https://github.com/git-for-windows/git/tree/HEAD/git-gui/lib and with a Windows-specific wrapper: https://github.com/git-for-windows/git/blob/HEAD/git-gui/windows/git-gui.sh

> the installer of git-for-windows comes with a `git-gui.exe` (not tcl-only)

Indeed, this is [_another_ wrapper](https://github.com/git-for-windows/git/wiki/Git-wrapper) that is [copied to `/cmd/git-gui.exe`](https://github.com/git-for-windows/MINGW-packages/blob/907ead42e93c073c0a37b902df02ef0d08102a99/mingw-w64-git/mingw-w64-git.mak#L24-L27).

> the [readme in the git-gui folder](https://github.com/git-for-windows/git/tree/main/git-gui#building-and-installing) mentions a "build" using `make`.

Right, and the idea was to replicate the same functionality in the CMake configuration, but that has not happened yet: https://github.com/git-for-windows/git/blob/f46b4dcad95e58cebbb9a9210342c1c56a43c567/contrib/buildsystems/CMakeLists.txt#L127

> Can someone help me start debugging of git-gui?

The easiest way is to use a PortableGit and edit `mingw64/libexec/git-core/git-gui.tcl` (which is a copy of `git-gui/git-gui.sh`) and the files in `mingw64/share/git-gui/lib/` (which are copies of the `git-gui/lib/` files) and _maybe_ even the small wrapper in `mingw64/libexec/git-core/git-gui.tcl` (which is a copy of `git-gui/windows/git-gui.sh`).

Once you implemented a new feature or a bug fix that way, copy the edited files over the source files and commit. (You may want to stage in a more fine-grained manner, e.g. using Git GUI itself, seeing as there are [some placeholders that are interpolated during a `make` run](https://github.com/git-for-windows/git/blob/f46b4dcad95e58cebbb9a9210342c1c56a43c567/git-gui/Makefile#L207-L214).)

**Reply by ToBoMi** [reply]
Thanks for your detailed response.
I was able to start developing in a copy of my existing git installation.

It is a bit strange to use an installation and not the source code, but since there is no other way I guess I have to live with it.
